### ISSUE TITLE
Python 3: error raised when moving into NVDA settings category

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#9768]
cc @michaeldcurran 

### Steps to reproduce:
From threshold_py3_staging:

1. Open NVDA"s settings dialog from the menu.
2. From the list of categories, press tab

### Actual behavior:
An error sound is heard.

<details>
<summary>Exception</summary>

```
ERROR - unhandled exception (20:03:12.984):
Traceback (most recent call last):
  File "baseObject.py", line 144, in _getPropertyViaCache
    val=self._propertyCache[getterMethod]
KeyError: <function Dialog._get_isPresentableFocusAncestor at 0x047F76A8>
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "baseObject.py", line 144, in _getPropertyViaCache
    val=self._propertyCache[getterMethod]
KeyError: <function IAccessible._get_isPresentableFocusAncestor at 0x04A28ED0>
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "baseObject.py", line 144, in _getPropertyViaCache
    val=self._propertyCache[getterMethod]
KeyError: <function NVDAObject._get_isPresentableFocusAncestor at 0x047D7810>
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "baseObject.py", line 144, in _getPropertyViaCache
    val=self._propertyCache[getterMethod]
KeyError: <function NVDAObject._get_presentationType at 0x047D7270>
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "baseObject.py", line 144, in _getPropertyViaCache
    val=self._propertyCache[getterMethod]
KeyError: <function Dialog._get_description at 0x047F7660>
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "baseObject.py", line 144, in _getPropertyViaCache
    val=self._propertyCache[getterMethod]
KeyError: <function IAccessible._get_description at 0x04A26F60>
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "D:\Development\nvda\include\comtypes\comtypes\_comobject.py", line 155, in call_without_this
    args[args_out_idx[i]][0] = value
  File "D:\Development\nvda\include\comtypes\comtypes\automation.py", line 575, in __setitem__
    self[index].value = value
  File "D:\Development\nvda\include\comtypes\comtypes\automation.py", line 218, in _set_value
    _VariantClear(self)
  File "_ctypes/callproc.c", line 933, in GetResult
OSError: [WinError -2147352568] Bad variable type
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 232, in 'calling callback function'
  File "D:\Development\nvda\include\comtypes\comtypes\_comobject.py", line 176, in call_without_this
    hr = HRESULT_FROM_WIN32(winerror(details))
  File "D:\Development\nvda\include\comtypes\comtypes\_comobject.py", line 51, in winerror
    code = exc[0]
TypeError: 'OSError' object is not subscriptable
```

</details>

I'm not very well accustomed to this new traceback presentation style, but the OSError one really bothers me and seems like a comtypes issue.

### Expected behavior:
NO errors

#### NVDA installed/portable/running from source:
Running from source

#### NVDA version:
At the time of writing, the try-py3_ignoreSomeIssues branch.

#### Windows version:
10 1903

**Comment by michaelDCurran** [comment]
So far I am guessing this is some kind of buffer overrun or other memory corruption.
It is specifically caused when calling accDescription(0) on the IAccessible for the propertyPage.
My guess is that the [out] VARIANT that is allocated by oleacc when calling GetPropValue on our AccPropServer is either not correctly initialized with VARIANTInit, or a buffer overrun has placed junk at that memory address after initialization of the variant and before comtypes assigns the return value from getPropValue. 
This is now the second strange bug we've seen with our accPropServer code. I think there is some mistake here... not sure what though.

**Comment by michaelDCurran** [comment]
I'm thinking the difference between Python2 and 3 here is simply just memory layout, and not a diliberet code change.

**Comment by michaelDCurran** [comment]
A part from the possible memory corruption, there is of course also an error in comtypes in regards to python3. 
Online 51 of _comobj.py,
```
code = exc[0]
```
Should be
```
code = exc.winerror
```
As WinError / WindowsError now throws OSError, which is unsubscritable now, plus the errno is now Operating System agnostic. Thus comtypes still requires the original windows error number. 
 I will file this with comtypes.

**Comment by feerrenrut** [comment]
> This is now the second strange bug we've seen with our accPropServer code. I think there is some mistake here... not sure what though.

While I think it only masks the issue, if you comment out the annotation for `ACC_DESCRIPTION` in the settings panel, the exception stops occurring.

This is quite similar to the issue I saw when trying to add an accessible description. While initially debugging that problem (acc description annotation not having any effect when attached to the window title), I found that upon copying the description there was an exception  / error in comtypes. 

>I will file this with comtypes.

Do you have link to the issue tracker for that?

**Comment by leonardder** [comment]
It looks like there is not yet an issue for comtypes.

**Comment by leonardder** [comment]
I filed https://github.com/enthought/comtypes/issues/186

**Comment by leonardder** [comment]
> Online 51 of _comobj.py,
> 
> ```
> code = exc[0]
> ```
> 
> Should be
> 
> ```
> code = exc.winerror
> ```

This is incorrect. winerror is the correct property for OSError, but comtypes._comobject.winerror is also said to support COMError, on which hresult is the proper attribute.

**Comment by leonardder** [comment]
Also filed https://github.com/enthought/comtypes/pull/187. Unfortunately, applying this pr will only obfuscate the exception, i.e. it no longer causes an error to be logged but yet IAccessible accDescription will still be None according to the dev info.

**Comment by leonardder** [comment]
I've changed the accPropServer GetPropValue method into a [low level implementation](https://pythonhosted.org/comtypes/server.html#low-level-implementation) but that doesn't improve anything. Things keep failing at _VariantClear, which seems to receive a malformed variant.

**Comment by leonardder** [comment]
@feerrenrut: I'm probably doing a last attempt to fix this either today or tomorrow before I"m out of ideas.
Given the other issues we have with property overrides on accprop servers which seem to be mostly related to BSTR in variants, I"m tempted to remove the description overriding altogether. Note that the NVDA dialog class has logic to calculate the description from static text children on the object, so basically, if we stop doing this with our accprop server, the user won't see a difference.
We can try to reimplement this as soon as WXPython 4.1 arrives, which should add support for IAccessible annotation using wx itself.

**Comment by feerrenrut** [comment]
If you apply the attached [patch.txt](https://github.com/nvaccess/nvda/files/3355657/patch.txt) in `include/comtypes` I believe the issue is resolved. I think this would likely fix the issue I was having with setting the name property as well.

I think this proves the theory that the `VARIANT` given as an out-param has not been initialised correctly, and there is a random value in the `VT` part. I think it's a bug in comtypes that it is trying to call `VariantClear` on this out-param at all.

So to summarise, I think this is the result of an intersection of mistakes.
- The windows api creating the `VARIANT` out-param should have initialised it.
- `comTypes` should never call `VariantClear` on out-params.

If you would like to put this issue down, let me know.

**Comment by leonardder** [comment]
Would you be able to throw this in a pr against comtypes, or should we add this to our monkeypatches?
We can build a new binary copy from one of our forks with the prs applied.

**Comment by feerrenrut** [comment]
Closed with commit a92ac90 in #9900
